src/main/java/com/facebook/presto/pinot/PinotScatterGatherQueryClient.java

```diff
@@ -0,0 +1,397 @@
+
+    public PinotScatterGatherQueryClient(Config pinotConfig)
+    {
+        prestoHostId = getDefaultPrestoId();
+
+        registry = new MetricsRegistry();
+        brokerMetrics = new BrokerMetrics(registry, DEFAULT_EMIT_TABLE_LEVEL_METRICS);
+        brokerMetrics.initializeGlobalMeters();
+
+        eventLoopGroup = new NioEventLoopGroup();
+        /**
+         * Some of the client metrics uses histogram which is doing synchronous operation.
+         * These are fixed overhead per request/response.
+         * TODO: Measure the overhead of this.
+         */
+        final NettyClientMetrics clientMetrics = new NettyClientMetrics(registry, "presto_pinot_client_");
+
+        // Setup Netty Connection Pool
+        resourceManager = new PooledNettyClientResourceManager(eventLoopGroup, new HashedWheelTimer(), clientMetrics);
+
+        requestSenderPool = Executors.newFixedThreadPool(pinotConfig.getThreadPoolSize());
+        poolTimeoutExecutor = new ScheduledThreadPoolExecutor(50);
+        connPool = new KeyedPoolImpl<>(pinotConfig.getMinConnectionsPerServer(), pinotConfig.getMaxConnectionsPerServer(), pinotConfig.getIdleTimeoutMs(),
+                pinotConfig.getMaxBacklogPerServer(), resourceManager, poolTimeoutExecutor, requestSenderPool, registry);
+        resourceManager.setPool(connPool);
+
+        // Setup ScatterGather
+        scatterGatherer = new ScatterGatherImpl(connPool, requestSenderPool);
+    }
+
+    public static <T> T doWithRetries(int numRetries, Function<Integer, T> caller)
+    {
+        PinotException firstError = null;
+        for (int i = 0; i < numRetries; ++i) {
+            try {
+                return caller.apply(i);
+            }
+            catch (PinotException e) {
+                if (firstError == null) {
+                    firstError = e;
+                }
+                if (!e.getErrorCode().isRetriable()) {
+                    throw e;
+                }
+            }
+        }
+        throw firstError;
+    }
+
+    private String getDefaultPrestoId()
+    {
+        String defaultBrokerId;
+        try {
+            defaultBrokerId = PRESTO_HOST_PREFIX + InetAddress.getLocalHost().getHostName();
+        }
+        catch (UnknownHostException e) {
+            defaultBrokerId = PRESTO_HOST_PREFIX;
+        }
+        return defaultBrokerId;
+    }
+
+    public Map<ServerInstance, DataTable> queryPinotServerForDataTable(String pql, String serverHost, List<String> segments, long connectionTimeoutInMs, boolean ignoreEmptyResponses, int pinotRetryCount)
+    {
+        BrokerRequest brokerRequest;
+        try {
+            brokerRequest = REQUEST_COMPILER.compileToBrokerRequest(pql);
+        }
+        catch (Pql2CompilationException e) {
+            throw new PinotException(ErrorCode.PINOT_INVALID_PQL_GENERATED, format("Parsing error with on %s, Error = %s", serverHost, e.getMessage()), e);
+        }
+
+        Map<String, List<String>> routingTable = new HashMap<>();
```

Comment:
Use guava's ImmutableMap.builder

Reply:
I don't want to bring in a dependency on guava here because then I would have to shade/exclude that etc. Instead I will make the caller (which is just one place in the presto-pinot code) wrap a ImmutableMap.copyOf() around this call.

Reply:
make sense